components/psst/psst_tab_helper.cc

```diff
@@ -0,0 +1,119 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/components/psst/psst_tab_helper.h"
+
+#include <string>
+#include <utility>
+
+#include "base/functional/bind.h"
+#include "base/memory/weak_ptr.h"
+#include "base/strings/utf_string_conversions.h"
+#include "base/task/thread_pool.h"
+#include "base/values.h"
+#include "brave/components/psst/psst_rule.h"
+#include "brave/components/psst/psst_service.h"
+#include "chrome/common/chrome_isolated_world_ids.h"
+#include "components/sessions/content/session_tab_helper.h"
+#include "content/public/browser/navigation_entry.h"
+#include "content/public/browser/navigation_handle.h"
+#include "content/public/browser/web_contents.h"
+
+namespace psst {
+
+PsstTabHelper::PsstTabHelper(content::WebContents* web_contents)
+    : WebContentsObserver(web_contents),
+      content::WebContentsUserData<PsstTabHelper>(*web_contents),
+      tab_id_(sessions::SessionTabHelper::IdForTab(web_contents)),
+      weak_factory_(this) {
+  if (!tab_id_.is_valid()) {
+    return;
+  }
+
+  psst_service_ = PsstService::GetInstance();
+  if (!psst_service_) {
+    return;
+  }
+}
+
+PsstTabHelper::~PsstTabHelper() = default;
+
+void InsertScriptInPage(
+    const content::GlobalRenderFrameHostId render_frame_host_id,
+    const std::string& script,
+    content::RenderFrameHost::JavaScriptResultCallback cb) {
+  content::RenderFrameHost* render_frame_host =
+      content::RenderFrameHost::FromID(render_frame_host_id);
+
+  if (render_frame_host) {
+    render_frame_host->ExecuteJavaScriptInIsolatedWorld(
+        base::UTF8ToUTF16(script), std::move(cb),
+        ISOLATED_WORLD_ID_BRAVE_INTERNAL);
+  }
+}
+
+void InsertPolicyScript(
+    const content::GlobalRenderFrameHostId render_frame_host_id,
+    std::string policy_script) {
+  InsertScriptInPage(render_frame_host_id, policy_script, base::DoNothing());
+}
+
+void OnTestScriptResult(
+    std::string policy_script,
+    const content::GlobalRenderFrameHostId render_frame_host_id,
+    base::Value value) {
+  if (!value.is_bool()) {
+    return;
+  }
+  const bool execute_policy = value.GetBool();
+  if (execute_policy) {
+    InsertPolicyScript(render_frame_host_id, policy_script);
+  }
+}
+
+void InsertTestScript(
+    const content::GlobalRenderFrameHostId render_frame_host_id,
+    MatchedRule rule) {
+  CHECK(render_frame_host_id);
+
+  InsertScriptInPage(render_frame_host_id, rule.test_script,
+                     base::BindOnce(&OnTestScriptResult, rule.policy_script,
+                                    render_frame_host_id));
+}
+
+void PsstTabHelper::DidFinishNavigation(
+    content::NavigationHandle* navigation_handle) {
+  CHECK(navigation_handle);
+
+  if (!psst_service_ || !navigation_handle->IsInPrimaryMainFrame() ||
+      !navigation_handle->HasCommitted() || !tab_id_.is_valid()) {
```

Comment:
this check seems redundant because `psst_service_` will be null if tab_id_ is not valid https://github.com/brave/brave-core/pull/20336/files#diff-c471942a0fd1ba00da849c3226c14aa64f7d772debc88e0a657948a1ceadc731R32